Improve death events

Phoenix616 opened the pull request:
This adds the ability to cancel the events and to specify the sound.

It should make handling of entity and player deaths a lot better than trying to fiddle with the damage event, especially as that has no ability to get the actual drops of an entity (or the proper death sound). There also is a method to set the health that the entity (or player) revives with (by default it will just use the maxhealth) which is also compatible with plugins that might set the health directly in the listener for whatever reason.

Phoenix616 committed:
Improve death events

This adds the ability to cancel the events and to specify the sound.

aikar commented on Spigot-Server-Patches/0356-Improve-death-events.patch:
```diff
@@ -0,0 +1,279 @@
++                }
++
++                this.getCombatTracker().g();
++                this.aU = true; // OBFHELPER - isDying
```
this isnt how the obf helpers work. you need to add a setter on this boolean, and label that of helper. then everything in thos code block thats obfuscated needs to be changed to obf helpers soo the mappings still line up.

we may want to comment out entire block above to force conflict though.

Phoenix616 replied:
Well I was just following other instances in the code where it was done this way but sure.

And by commenting out you mean using single line comments instead of the multi-line comment on lines 80 to 88 I guess?

aikar replied:
yes, single line comments, and that style your referencing is PAIL style from Spigot. They do that to mark 'rename this in the future'.
We want all multi line additions to be unbobfuscated so it's clear what the intended method/property to access/modify is.

With this this.aU, if the isDying field changes, it doesn't tell us where that boolean is in the class / inheritance chain. but if it conflicts on the actual property line, it's very clear, and then every where that property is used by paper is fixed.

Same for that .g() above it too, since its 'copied' logic, needs to be clear what the intent is.

however, I mention single line comments so that if mojang adds or removes lines in this body of code, we can detect that. /* style comments would open us up to desync.

Phoenix616 replied:
Yeah, makes sense, I'll change it to that,

aikar commented on Spigot-Server-Patches/0356-Improve-death-events.patch:
```diff
@@ -0,0 +1,279 @@
+ 
++    // Paper start - cancellable death event
++    public static CraftSound getBySoundEffect(final SoundEffect effect) {
++        MinecraftKey key = SoundEffect.a.b(effect);
```
need obf helper here

Phoenix616 committed:
Revive, not review...

Phoenix616 committed:
Add obfuscation helpers and fix player death cancelling

Also add a decompilation fix

aikar commented on Spigot-Server-Patches/0356-Improve-death-events.patch:
```diff
@@ -0,0 +1,616 @@
+ 
+     public void trackDamage(DamageSource damagesource, float f, float f1) {
+-        this.g();
++        this.reset(); // Paper
```
this didnt need to be changed, we only need to change lines we are touching.

aikar commented on Spigot-Server-Patches/0356-Improve-death-events.patch:
```diff
@@ -0,0 +1,616 @@
+             ExplosionPrimeEvent event = new ExplosionPrimeEvent(this.getBukkitEntity(), this.explosionRadius * f, false);
+             this.world.getServer().getPluginManager().callEvent(event);
+             if (!event.isCancelled()) {
+-                this.aU = true;
```
this didnt need to be changed, we only need to change lines we are touching.

aikar commented on Spigot-Server-Patches/0356-Improve-death-events.patch:
```diff
@@ -0,0 +1,616 @@
+         super.n();
+         this.world.methodProfiler.a("looting");
+-        if (!this.world.isClientSide && this.cX() && !this.aU && this.world.getGameRules().getBoolean("mobGriefing")) {
++        if (!this.world.isClientSide && this.cX() && !this.isDying() && this.world.getGameRules().getBoolean("mobGriefing")) { // Paper
```
this didnt need to be changed, we only need to change lines we are touching.

aikar commented on Spigot-Server-Patches/0356-Improve-death-events.patch:
```diff
@@ -0,0 +1,616 @@
+ 
+             if (this.ticksLived % 20 == 0) {
+-                this.getCombatTracker().g();
++                this.getCombatTracker().reset(); // Paper
```
this didnt need to be changed, we only need to change lines we are touching.

aikar commented on Spigot-Server-Patches/0356-Improve-death-events.patch:
```diff
@@ -0,0 +1,616 @@
+             }
+ 
+-            entityliving.a(this, this.bb, damagesource);
++            entityliving.runKillTrigger(this, this.getKillCount(), damagesource); // Paper
```
this didnt need to be changed, we only need to change lines we are touching.

aikar commented on Spigot-Server-Patches/0356-Improve-death-events.patch:
```diff
@@ -0,0 +1,616 @@
+         this.extinguish();
+         this.setFlag(0, false);
+-        this.getCombatTracker().g();
++        this.getCombatTracker().reset(); // Paper
```
this didnt need to be changed, we only need to change lines we are touching.

aikar commented on Spigot-Server-Patches/0356-Improve-death-events.patch:
```diff
@@ -0,0 +1,616 @@
+     @Nullable
+     public static MinecraftKey getName(Class<? extends Entity> oclass) {
+-        return (MinecraftKey) EntityTypes.b.b(oclass);
++        return (MinecraftKey) EntityTypes.b.getByValue(oclass); // Paper
```
this didnt need to be changed, we only need to change lines we are touching.

aikar commented on Spigot-Server-Patches/0356-Improve-death-events.patch:
```diff
@@ -0,0 +1,616 @@
+ 
+     public NBTTagCompound save(NBTTagCompound nbttagcompound) {
+-        MinecraftKey minecraftkey = (MinecraftKey) Item.REGISTRY.b(this.item);
++        MinecraftKey minecraftkey = (MinecraftKey) Item.REGISTRY.getByValue(this.item); // Paper
```
this didnt need to be changed, we only need to change lines we are touching.

aikar commented on Spigot-Server-Patches/0356-Improve-death-events.patch:
```diff
@@ -0,0 +1,616 @@
+         // CraftBukkit start - introduce a temporary timeout hack until this is fixed properly
+         if ((this.a.ticksLived - this.a.hurtTimestamp) > 100) {
+-            this.a.b((EntityLiving) null);
++            this.a.onKill((EntityLiving) null); // Paper
```
this didnt need to be changed, we only need to change lines we are touching.

aikar commented on Spigot-Server-Patches/0356-Improve-death-events.patch:
```diff
@@ -0,0 +1,616 @@
+ 
+             // Paper start - ignore missing recipes
+-            MinecraftKey key = CraftingManager.recipes.b(irecipe);
++            MinecraftKey key = CraftingManager.recipes.getByValue(irecipe); // Paper
```
this didnt need to be changed, we only need to change lines we are touching.

aikar commented on Spigot-Server-Patches/0356-Improve-death-events.patch:
```diff
@@ -0,0 +1,616 @@
+             IRecipe irecipe = (IRecipe) iterator1.next();
+ 
+-            MinecraftKey key = CraftingManager.recipes.b(irecipe);
++            MinecraftKey key = CraftingManager.recipes.getByValue(irecipe); // Paper
```
this didnt need to be changed, we only need to change lines we are touching.

aikar commented on Spigot-Server-Patches/0356-Improve-death-events.patch:
```diff
@@ -0,0 +1,616 @@
+         MinecraftKey minecraftkey = new MinecraftKey(s);
+ 
+-        SoundEffect.a.a(SoundEffect.c++, minecraftkey, new SoundEffect(minecraftkey));
++        SoundEffect.getRegistry().a(SoundEffect.c++, minecraftkey, new SoundEffect(minecraftkey)); // Paper
```
this didnt need to be changed, we only need to change lines we are touching.

aikar commented on Spigot-Server-Patches/0356-Improve-death-events.patch:
```diff
@@ -0,0 +1,616 @@
+     @Nullable public static MinecraftKey getKey(Class<? extends TileEntity> oclass) { return a(oclass); } // Paper - OBFHELPER
+     @Nullable public static MinecraftKey a(Class<? extends TileEntity> oclass) {
+-        return (MinecraftKey) TileEntity.f.b(oclass);
++        return (MinecraftKey) TileEntity.f.getByValue(oclass); // Paper
```
this didnt need to be changed, we only need to change lines we are touching.

aikar commented on Spigot-Server-Patches/0356-Improve-death-events.patch:
```diff
@@ -0,0 +1,616 @@
+ 
+     private NBTTagCompound c(NBTTagCompound nbttagcompound) {
+-        MinecraftKey minecraftkey = (MinecraftKey) TileEntity.f.b(this.getClass());
++        MinecraftKey minecraftkey = (MinecraftKey) TileEntity.f.getByValue(this.getClass()); // Paper
```
this didnt need to be changed, we only need to change lines we are touching.

aikar commented on Spigot-Server-Patches/0356-Improve-death-events.patch:
```diff
@@ -0,0 +1,616 @@
+         crashreportsystemdetails.a("Name", new CrashReportCallable() {
+             public String a() throws Exception {
+-                return TileEntity.f.b(TileEntity.this.getClass()) + " // " + TileEntity.this.getClass().getCanonicalName();
++                return TileEntity.f.getByValue(TileEntity.this.getClass()) + " // " + TileEntity.this.getClass().getCanonicalName(); // Paper
```
this didnt need to be changed, we only need to change lines we are touching.

aikar commented on Spigot-Server-Patches/0356-Improve-death-events.patch:
```diff
@@ -0,0 +1,616 @@
+ 
+     public static SoundEffect getSoundEffect(String s) {
+-        SoundEffect effect = SoundEffect.a.get(new MinecraftKey(s));
++        SoundEffect effect = SoundEffect.getRegistry().get(new MinecraftKey(s)); // Paper
```
this didnt need to be changed, we only need to change lines we are touching.

aikar commented on Spigot-Server-Patches/0356-Improve-death-events.patch:
```diff
@@ -0,0 +1,616 @@
+         }
+ 
+-        return Biome.valueOf(BiomeBase.REGISTRY_ID.b(base).getKey().toUpperCase(java.util.Locale.ENGLISH));
++        return Biome.valueOf(BiomeBase.REGISTRY_ID.getByValue(base).getKey().toUpperCase(java.util.Locale.ENGLISH)); // Paper
```
this didnt need to be changed, we only need to change lines we are touching.

aikar commented on Spigot-Server-Patches/0356-Improve-death-events.patch:
```diff
@@ -0,0 +1,616 @@
+     @Test
+     public void testReverse() {
+-        for (MinecraftKey effect : SoundEffect.a.keySet()) {
++        for (MinecraftKey effect : SoundEffect.getRegistry().keySet()) { // Paper
```
this didnt need to be changed, we only need to change lines we are touching.

aikar reviewed (approved):
accidently clicked approve, see requested changes

aikar reviewed (changes requested):
we don't unobfuscate lines that arent part of our change set, as that adds massive patch burden to us.

please remove the changes on each line i commented on

Phoenix616 committed:
Remove unnecessary deobfuscation

Phoenix616 committed:
Directly set the max health of the entity in the event
Also use double instead of float for the health like the API does everywhere else

Phoenix616 committed:
Fix issues with double-float conversions
The actual attributes are doubles but the internal getMaxHealth method returns a float. That was getting converted back to a double when checking the range in the setReviveHealth method which could lead to it being bigger than the actual max health attribute. This fixes that issue.
